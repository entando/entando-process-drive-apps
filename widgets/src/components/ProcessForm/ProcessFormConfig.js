import React from 'react';
import PropTypes from 'prop-types';
import i18next from 'i18next';
import { FormGroup, ControlLabel, HelpBlock, Row, Col } from 'patternfly-react';

import { getProcesses } from 'api/pda/processes';
import { getConnections } from 'api/pda/connections';
import JsonMultiFieldContainer from 'components/common/form/SchemaEditor/JsonMultiFieldContainer';

class ProcessFormConfig extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: false,

      sourceList: [],
      processList: [],
      config: {
        settings: {
          uiSchemas: '[]',
          processDefinition: '',
        },
        knowledgeSource: '',
      },
    };

    this.fetchOnLoad = this.fetchOnLoad.bind(this);
    this.fetchFromKnowledgeSource = this.fetchFromKnowledgeSource.bind(this);

    this.onChangeKnowledgeSource = this.onChangeKnowledgeSource.bind(this);
    this.onChangeProcessDefinition = this.onChangeProcessDefinition.bind(this);
    this.onChangeUiSchemas = this.onChangeUiSchemas.bind(this);
  }

  componentDidMount() {
    this.fetchOnLoad();
  }

  componentDidUpdate(prevProps) {
    const { config } = this.props;

    // refill state if passed (props) config changes
    if (JSON.stringify(config) !== JSON.stringify(prevProps.config)) {
      this.fetchOnLoad();
    }
  }

  async onChangeKnowledgeSource({ target: { value } }) {
    const { config } = this.state;

    this.setState(
      {
        config: {
          ...config,
          knowledgeSource: value,
        },
      },
      async () => {
        const processList = await this.fetchProcess();
        this.setState({ processList });
      }
    );
  }

  onChangeProcessDefinition({ target: { value } }) {
    const { config } = this.state;

    this.setState({
      config: {
        ...config,
        settings: {
          ...config.settings,
          processDefinition: value,
        },
      },
    });
  }

  onChangeUiSchemas(uiSchemas) {
    const { config } = this.state;
    this.setState({
      config: {
        ...config,
        settings: { ...config.settings, uiSchemas },
      },
    });
  }

  fetchProcess = async () => {
    const { config } = this.state;

    const connection = (config && config.knowledgeSource) || '';

    if (connection.length) {
      try {
        const processList = await getProcesses(connection);
        const { payload, errors } = processList;
        if (errors && errors.length) {
          throw errors[0];
        }
        return payload.map(proc => ({
          label: proc['process-name'],
          value: `${proc['process-id']}@${proc['container-id']}`,
        }));
      } catch (error) {
        this.handleError(error.message);
      }
    }
    return [];
  };

  fetchOnLoad() {
    const { config } = this.props;

    // getting list of Kie server connections
    this.setState({ loading: true }, async () => {
      const {
        sourceList = [],
        selectedKnowledgeSource = '',
      } = await this.fetchFromKnowledgeSource();

      const parsedSettings = config.settings ? JSON.parse(config.settings) : {};

      this.setState(
        {
          loading: false,
          sourceList,
          processList: [],
          config: {
            ...config,
            settings: parsedSettings,
            knowledgeSource: selectedKnowledgeSource,
          },
        },
        async () => {
          const processList = await this.fetchProcess();
          this.setState({ processList });
        }
      );
    });
  }

  async fetchFromKnowledgeSource() {
    const {
      config: { knowledgeSource },
    } = this.props;

    const { payload: sourceList } = await getConnections();

    // checking if connection (knowledgeSource) was previously selected and exists in the
    // list of available connections - processes are defined per connection
    const isSelectable =
      knowledgeSource &&
      sourceList.some(iteratedConnection => iteratedConnection.name === knowledgeSource);

    // fetching other values (like process) that depend on connection (knowledgeSource)
    // could be done here by further chaining them (for an example, refer to ProcessForm)

    // if previously selected connection (knowledgeSource) is not available, it should be deselected
    return {
      sourceList,
      ...(isSelectable
        ? { selectedKnowledgeSource: knowledgeSource }
        : { selectedKnowledgeSource: '' }),
    };
  }

  render() {
    const { config, sourceList, loading, processList } = this.state;
    const { knowledgeSource, settings } = config;

    return (
      <div>
        <form>
          <Row>
            <Col xs={12}>
              <FormGroup controlId="connection">
                <ControlLabel>{i18next.t('config.knowledgeSource')}</ControlLabel>
                <select
                  disabled={loading}
                  className="form-control"
                  value={knowledgeSource}
                  onChange={this.onChangeKnowledgeSource}
                >
                  <option disabled value="">
                    {i18next.t('config.selectOption')}
                  </option>
                  {sourceList.map(source => (
                    <option key={source.name} value={source.name}>
                      {source.name}
                    </option>
                  ))}
                </select>
                <HelpBlock>{i18next.t('config.selectConnections')}</HelpBlock>
              </FormGroup>
            </Col>
          </Row>
          {knowledgeSource && (
            <section>
              <legend>{i18next.t('config.settings')}</legend>
              <Row>
                <Col xs={12}>
                  <FormGroup bsClass="form-group" controlId="textarea">
                    <ControlLabel bsClass="control-label">
                      {i18next.t('config.processDefinition')}
                    </ControlLabel>
                    <select
                      disabled={loading}
                      className="form-control"
                      value={settings.processDefinition}
                      onChange={this.onChangeProcessDefinition}
                    >
                      <option disabled value="">
                        {i18next.t('config.selectOption')}
                      </option>
                      {processList.map(process => (
                        <option key={process.value} value={process.value}>
                          {process.label}
                        </option>
                      ))}
                    </select>
                  </FormGroup>
                </Col>
              </Row>
              <Row>
                <Col xs={12}>
                  <FormGroup bsClass="form-group" controlId="textarea">
                    <ControlLabel bsClass="control-label">
                      {i18next.t('config.uiSchemas')}
                    </ControlLabel>
                    <JsonMultiFieldContainer
                      schemas={settings.uiSchemas}
                      onChange={this.onChangeUiSchemas}
                    />
                  </FormGroup>
                </Col>
              </Row>
            </section>
          )}
        </form>
      </div>
    );
  }
}

ProcessFormConfig.propTypes = {
  config: PropTypes.shape({
    knowledgeSource: PropTypes.string,
    settings: PropTypes.string,
  }),
};

ProcessFormConfig.defaultProps = {
  config: {},
};

export default ProcessFormConfig;
