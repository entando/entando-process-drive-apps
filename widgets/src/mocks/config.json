{
  "payload": {
    "code": "pda_task_list",
    "config": {
      "groups": "[{\"label\":\"Administrators\",\"key\":\"Administrators\",\"checked\":true},{\"label\":\"broker\",\"key\":\"broker\",\"checked\":true},{\"label\":\"admin\",\"key\":\"admin\",\"checked\":false},{\"label\":\"rest-all\",\"key\":\"rest-all\",\"checked\":false},{\"label\":\"kiemgmt\",\"key\":\"kiemgmt\",\"checked\":true},{\"label\":\"kie-server\",\"key\":\"kie-server\",\"checked\":true},{\"label\":\"appraiser\",\"key\":\"appraiser\",\"checked\":false},{\"label\":\"manager\",\"key\":\"manager\",\"checked\":true},{\"label\":\"supplier\",\"key\":\"supplier\",\"checked\":true},{\"label\":\"IT\",\"key\":\"IT\",\"checked\":true},{\"label\":\"PM\",\"key\":\"PM\",\"checked\":true}]",
      "options": "[{\"key\":\"newPageOnClick\",\"label\":\"Open new page on table row click\",\"checked\":false},{\"key\":\"showClaim\",\"label\":\"Show Claim Button\",\"checked\":true},{\"key\":\"showComplete\",\"label\":\"Show Complete Button\",\"checked\":true}]",
      "columns": "[{\"name\":\"task-is-skipable\",\"position\":0,\"isVisible\":false},{\"name\":\"task-parent-id\",\"position\":1,\"isVisible\":false},{\"name\":\"task-created-by\",\"position\":2,\"isVisible\":true},{\"name\":\"task-input-data.application.property.saleprice\",\"position\":3,\"isVisible\":true},{\"name\":\"task-input-data.TaskName\",\"position\":4,\"isVisible\":true},{\"name\":\"initiator\",\"position\":5,\"isVisible\":true},{\"name\":\"task-created-on\",\"position\":6,\"isVisible\":true},{\"name\":\"task-priority\",\"position\":7,\"isVisible\":false},{\"name\":\"task-proc-def-id\",\"position\":8,\"isVisible\":true},{\"name\":\"task-form\",\"position\":9,\"isVisible\":true},{\"name\":\"task-skippable\",\"position\":10,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.name\",\"position\":11,\"isVisible\":true},{\"name\":\"task-actual-owner\",\"position\":12,\"isVisible\":true},{\"name\":\"task-activation-time\",\"position\":13,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.address\",\"position\":14,\"isVisible\":true},{\"name\":\"task-input-data.application.errors\",\"position\":15,\"isVisible\":true},{\"name\":\"task-name\",\"position\":16,\"isVisible\":true},{\"name\":\"task-input-data.application.property.address\",\"position\":17,\"isVisible\":true},{\"name\":\"task-subject\",\"position\":18,\"isVisible\":true},{\"name\":\"containerId\",\"position\":19,\"isVisible\":true},{\"name\":\"task-container-id\",\"position\":20,\"isVisible\":true},{\"name\":\"processDefinitionId\",\"position\":21,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.ssn\",\"position\":22,\"isVisible\":true},{\"name\":\"task-description\",\"position\":23,\"isVisible\":true},{\"name\":\"task-input-data.Skippable\",\"position\":24,\"isVisible\":true},{\"name\":\"task-workitem-id\",\"position\":25,\"isVisible\":true},{\"name\":\"task-expiration-time\",\"position\":26,\"isVisible\":true},{\"name\":\"task-input-data.application.property.age\",\"position\":27,\"isVisible\":true},{\"name\":\"task-status\",\"position\":28,\"isVisible\":true},{\"name\":\"task-input-data.application.mortgageamount\",\"position\":29,\"isVisible\":true},{\"name\":\"task-process-instance-id\",\"position\":30,\"isVisible\":true},{\"name\":\"task-input-data.ActorId\",\"position\":31,\"isVisible\":true},{\"name\":\"task-process-id\",\"position\":32,\"isVisible\":true},{\"name\":\"task-input-data.application.property.locale\",\"position\":33,\"isVisible\":true},{\"name\":\"task-input-data.application.amortization\",\"position\":34,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.creditrating\",\"position\":35,\"isVisible\":true},{\"name\":\"application\",\"position\":36,\"isVisible\":true},{\"name\":\"task-input-data.NodeName\",\"position\":37,\"isVisible\":true},{\"name\":\"task-proc-inst-id\",\"position\":38,\"isVisible\":true},{\"name\":\"task-input-data.application.downpayment\",\"position\":39,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.annualincome\",\"position\":40,\"isVisible\":true},{\"name\":\"task-id\",\"position\":41,\"isVisible\":true}]",
      "knowledgeSource": "kieStaging",
      "process": "itorders.orderhardware@itorders_1.0.0-SNAPSHOT"
    }
  },
  "metaData": {
    "status": "draft"
  },
  "errors": []
}
