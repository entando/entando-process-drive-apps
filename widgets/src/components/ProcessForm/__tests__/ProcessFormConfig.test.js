import { render, wait } from '@testing-library/react';
import React from 'react';
import 'mocks/i18nMock';

import WIDGET_CONFIGS from 'mocks/app-builder/pages';
import CONNECTIONS from 'mocks/pda/connections';
import PROCESSES from 'mocks/pda/processes';
import ProcessFormConfig from 'components/ProcessForm/ProcessFormConfig';

describe('<ProcessFormConfig />', () => {
  it('renders snapshot correctly', async () => {
    fetch
      .once(JSON.stringify(CONNECTIONS))
      .once(JSON.stringify(WIDGET_CONFIGS.PROCESS_FORM))
      .once(JSON.stringify(PROCESSES));

    const { container } = render(
      <ProcessFormConfig pageCode="phase_1_widgets" frameId="1" widgetCode="process_form" />
    );

    await wait(() => expect(container).toMatchSnapshot());
  });
});
