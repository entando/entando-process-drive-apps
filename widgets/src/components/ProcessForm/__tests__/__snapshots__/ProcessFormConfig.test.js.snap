// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProcessFormConfig /> renders snapshot correctly 1`] = `
<div>
  <div>
    <form>
      <div
        class="row"
      >
        <div
          class="col-xs-12"
        >
          <div
            class="form-group"
          >
            <label
              class="control-label"
              for="connection"
            >
              config.knowledgeSource
            </label>
            <select
              class="form-control"
            >
              <option
                disabled=""
                value=""
              >
                config.selectOption
              </option>
              <option
                value="kieStaging"
              >
                kieStaging
              </option>
              <option
                value="kie2Server"
              >
                kie2Server
              </option>
              <option
                value="test"
              >
                test
              </option>
            </select>
            <span
              class="help-block"
            >
              config.selectConnections
            </span>
          </div>
        </div>
      </div>
      
    </form>
  </div>
</div>
`;
